bento/src/main/java/com/yelp/android/bento/core/AsyncInflationBridge.kt

```diff
@@ -116,13 +114,150 @@ internal class AsyncInflationBridge @JvmOverloads constructor(
      */
     fun trackComponentRemoval(component: Component) {
-        inflationJobs[component]?.cancel()
+        inflationJobs.remove(component)?.cancel()
     }
 
     fun cancelAllInflationJobs() {
         clearResources()
     }
 
+    /**
+     * Intercepts addComponent() calls from a RecyclerViewComponentController and pre-inflates
+     * views asynchronously based on the set [AsyncInflationStrategy].
+     */
+    fun asyncInflateViewsForComponent(component: Component, addComponentCallback: () -> Unit) {
+        if (strategy == DEFAULT) {
+            if (smartAsyncCacheEmptyOnStart) {
+                asyncInflateViewsBestGuess(component, addComponentCallback)
+            } else {
+                smartAsyncInflateViewsForComponent(addComponentCallback)
+            }
+        } else if (strategy == SMART) {
+            smartAsyncInflateViewsForComponent(addComponentCallback)
+        } else if (strategy == BEST_GUESS) {
+            asyncInflateViewsBestGuess(component, addComponentCallback)
+        }
+    }
+
+    /**
+     * The "SMART" strategy for async inflation where adding components blocks on inflating above
+     * the fold components before inflating below the fold components.
+     */
+    private fun smartAsyncInflateViewsForComponent(addComponentCallback: () -> Unit) {
+        var aboveFold: List<Any> = mutableListOf()
+        launch(defaultBridgeDispatcher) {
+            // The lock will block all addComponent calls that occur until the above the fold
+            // view holders are inflated.
+            lock.withLock {
+                // Only inflate above the fold if the cache wasn't empty in create, the below the
+                // fold set hasn't been inflated yet and if there's a match for this page name in
+                // the cache.
+                if (!belowTheFoldTriggered && !smartAsyncCacheEmptyOnStart &&
+                        SmartAsyncInflationCache.containsKey(name)) {
+                    val mapForThisView = SmartAsyncInflationCache[name] ?: mutableMapOf()
+                    var count = numberOfAboveTheFoldComponents
+                    aboveFold = mapForThisView.keys.takeWhile { count--; count > -1 }
+                    if (!aboveTheFoldTriggered) {
+                        asyncInflateViewsIfPossible(aboveFold, true)
+                        aboveTheFoldTriggered = true
+                    }
+                }
+                withContext(Dispatchers.Main) {
+                    executeAddComponentCallback(addComponentCallback)
+                }
+            }
+            // Outside the lock. The second time it's called, it should inflate the below the fold
+            // view holders without blocking any more addComponent calls.
+            if (!belowTheFoldTriggered && !smartAsyncCacheEmptyOnStart &&
+                    SmartAsyncInflationCache.containsKey(name)) {
+                asyncInflateViewsIfPossible(aboveFold, false)
+            }
```

Comment:
This part confuses me, the fact that we can call async inflate in some case after the callback has been executed.
I will need to re-read that tomorrow.